java/change-notes/2021-05-14-close-resource-leaks-improvements.md

```diff
@@ -0,0 +1,2 @@
+lgtm,codescanning
+* The "Potential input resource leak" (`java/input-resource-leak`) and "Potential output resource leak" (`java/output-resource-leak`) queries have been made more specific to only consider JDK classes and subtypes. Additionally the number of false positives has been reduced.
```

Comment:
```suggestion
* The "Potential input resource leak" (`java/input-resource-leak`) and "Potential output resource leak" (`java/output-resource-leak`) queries now produce less false negatives, because they only exclude some JDK classes and subtypes from consideration, not types that happen to have the same name as was previously the case. Additionally the number of false positives has been reduced by recognising `CharArrayReader` and `CharArrayWriter` as types that don't need to be closed.
```

Reply:
> now produce less false negatives, because they only exclude some JDK classes and subtypes from consideration

My intention with the change note was a different one, the changes might actually reduce true and false positives for classes with names such as `Reader`, `InputStream`, ... which are not the JDK classes. Since it affects true and false positives I only wrote that the query is now "more specific", without stating what effect that could have. But you are right, it would also affect false negatives, though maybe "more specific" suffices to cover both cases?

If you want I can explicitly mention `CharArrayReader` and `CharArrayWriter`, though I tried to prevent the change note from becoming too verbose.

Reply:
Updated my suggestion to be a bit particular about what we've fixed.

Reply:
That sounds good, thanks! Really minor, but "recogni<del>s</del><ins>z</ins>ing" (American English spelling), right?
Feel free to directly commit your suggestion on this branch then.